Two Twelve: Support building on 2.12.

Shadowfiend opened the pull request:
For your consideration, @farmdawgnation .

farmdawgnation reviewed (approved):
Pushed one minor change to actually add 2.12 to the build definition. Other than that, it's looking good! :)

Shadowfiend commented:
Cool. I need to figure out the 2.6 lift cross-build too <_<

Shadowfiend committed:
Rework how we represent HNil.

Scala 2.12 has some tweaked exhaustiveness checks on pattern matching that
catch a possibility we hadn't aniticipated. In particular, HNil was defined as
a final class, and we created a singleton instance of it to reference, but a
user could instantiate the final class themselves and send that in and wreak
havoc.

We redefine HNil as a final singleton object, and define a type HNil that is
equivalent to the singleton's type so we can reference it as both a value and
type called HNil. Other code remains unchanged, fortunately.

Unfortunately, this means we have to keep using our forkity fork of Lift's
HLists... Hopefully a later Lift minor version can include these tweaks, but
we'll need to maintain our fork until we no longer support versions of Lift
with the old implementation.

Shadowfiend committed:
Explicitly partially apply handleSubmit.

Scala 2.12 yells at us if we don't do it, so let's fix it!

Shadowfiend committed:
Bump to Scala 2.12.

Also bump to specs2 4.0.2, since 3.8 isn't built for Scala 2.12 apparently.

Shadowfiend committed:
Bump to sbt 1.1.0.

This requires reworking a few of our old expressions in the build file, but
nothing should be changing from a practical standpoint.

Shadowfiend committed:
Add Scala 2.12.4 to the Travis cross-compile.

Shadowfiend commented:
So I actually think this is good to go… Would be nice to add snapshot publishing on merge to master though…

Shadowfiend committed:
Auto-publish snapshots

We add a travis.sh build script that checks for this to be a non-PR build. The
build is only run on the master branch as a bonus. If all that is true AND we
are building a snapshot version, we descript the encrypted Sonatype credentials
and use them to publish the snapshot to Sonatype.

Shadowfiend committed:
Allow cross-builds to Scala 2.11

Shadowfiend committed:
Cross-build Lift versions

We build Lift 3.0, 3.1, 3.2 against Scala 2.11 and Scala 2.12, and we build
Lift 2.6 against Scala 2.11.

Shadowfiend committed:
Restore Scaladoc building

Since we use Java 8 in travis, we can return to publishing Scaladocs!

Shadowfiend committed:
Only publish snapshots for master branch

Shadowfiend commented:
Ok, pending some travis love, we should be in business here…

We now cross-build against Scala 2.11 and 2.12 for Lift versions 3.0, 3.1, and 3.2; we also build against Scala 2.11 for Lift version 2.6! The patch versions are the latest across the major/minor versions.

We also auto-publish on the master branch for maximum joy. I say that, but we'll see if it works in practice once the merge happens………

farmdawgnation commented on .travis.yml:
```diff
@@ -3,11 +3,25 @@ jdk:
+  - LIFT_VERSION=3.1.1
+  - LIFT_VERSION=3.0.2
+
+# Add a spritz of 2.6 build.
```
![just a spritz](https://media.giphy.com/media/X91oeDLYRhHEs/giphy.gif)

farmdawgnation commented on travis.sh:
```diff
@@ -0,0 +1,15 @@
+  sbt ++$TRAVIS_SCALA_VERSION "set liftVersion := \"$LIFT_VERSION\"" publish
+
+  rm ~/.sonatype
+fi
```
Missing newline to end the file

farmdawgnation reviewed (approved):
Picked a nit, but otherwise good to go. :+1:

Shadowfiend committed:
Newline nit

Shadowfiend commented:
Fix't, merging on green light!
